KAFKA-10357: Add explicit internal topic initialization for Kafka Streams

k-apol opened the pull request:
This PR implements the init() method described in KIP-698 to allow users
to validate or prepare internal Kafka Streams topics before starting the
application. The changes include:

Introduced a public init() method on KafkaStreams, which performs
internal topic validation and optionally sets up missing topics.

Added support for a new configuration option, internal.topics.setup,
with 'automatic' and 'manual' as options to control default internal
topic management.

Updated InternalTopicManager to expose validate, and allow for a
separate initialization timeout.

k-apol committed:
test commit

k-apol committed:
Add new internal topic setup config & related test

k-apol committed:
Update new internal topic doc to be private member

k-apol committed:
remove redundant lines in test case

k-apol committed:
Add boilerplate stub for init method and Javadoc comments

k-apol committed:
Add MissingInternalTopicsException per KIP

k-apol committed:
Add InternalTopicAlreadySetup exception

k-apol committed:
Add MisconfiguredInternalTopicException

k-apol committed:
Add boilerplate class for Initparameters

k-apol committed:
Add boilerplate init methods in KafkaStreams to support manual initialization + TODOs

k-apol committed:
new setup, test commit to verify connection to IDE

k-apol committed:
Add init method to Kafka Streams

k-apol committed:
Style fixes
Add timeout handler method to internalTopicManager to handle both init timeouts and retry timeouts
Cleanup ChangelogTopics

k-apol committed:
Add unit tests for init method

k-apol committed:
fix typo

k-apol committed:
typo

k-apol committed:
typo

k-apol committed:
Merge branch 'trunk' into KAFKA-10357-Accidental-Repartition-Handling

mjsax commented:
Seems there is some style errors: `Execution failed for task ':streams:spotlessJavaCheck'.`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -173,6 +173,7 @@ public class StreamsConfig extends AbstractConfig {
 
     public static final long MAX_TASK_IDLE_MS_DISABLED = -1;
 
+    public static final Duration DEFAULT_INIT_TIMEOUT_MS = Duration.ofMillis(30000);
```
I don't think we should add this here, as it would "leak" into public API. Not even sure if we need a variable for it at all?

k-apol replied:
I am marking this down along with the other concerns surrounding the timeout

mjsax commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -293,6 +302,168 @@ public boolean isValidTransition(final State newState) {
+     * @throws MisconfiguredInternalTopicException if an internal topics is misconfigured
+     * @throws InternalTopicsAlreadySetupException if all internal topics are already setup
+     */
+    public void init() {
```
Might also throw `TimeoutException`, right, as it applied the default timeout?

k-apol replied:
Thank you, added this and resolving

mjsax commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -293,6 +302,168 @@ public boolean isValidTransition(final State newState) {
+     * @throws InternalTopicsAlreadySetupException if all internal topics are already setup
+     */
+    public void init() {
+        this.init(DEFAULT_INIT_TIMEOUT_MS);
```
nit: avoid unnecessary `this.` prefix

k-apol replied:
Thank you, added this and resolving

mjsax commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -293,6 +302,168 @@ public boolean isValidTransition(final State newState) {
     protected volatile State state = State.CREATED;
 
+    /**
+     * Initializes broker-side state.
```
This description is a little bit brief. Would be good to add more context that "broker-side state" actually is, ect.

k-apol replied:
I am pushing an updated one, let me know what you think

mjsax commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -293,6 +302,168 @@ public boolean isValidTransition(final State newState) {
+            return setupInternalTopicsIfIncomplete;
+        }
+
+        public final void enableTimeout() {
```
This method was not defined on the KIP -- so we should not add it -- or we need to update the KIP.

Looking into KIP-1092 and KIP-1153 which introduce `CloseOptions` which also takes a timeout, it might make sense to update the KIP to include the timeout in `InitParameters` and remove the `Duration timeout` parameter and corresponding `init()` overloads? @cadonna WDYT?

k-apol replied:
Makes sense, I will update as this and the other timeout-related code as conversation evolves. There may be a better way to approach handling a timeout here.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -293,6 +302,168 @@ public boolean isValidTransition(final State newState) {
+            return new InitParameters(false);
+        }
+
+        public boolean setupInternalTopicsIfIncompleteEnabled() {
```
While the KIP does specify this getter, I am wondering why? We usually don't use getters on the main config-classes, but use sub-classes. Might be good to re-read the KIP discussion (and if need be, update the KIP to "fix" this)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -293,6 +302,168 @@ public boolean isValidTransition(final State newState) {
+                allSourceTopics.addAll(topicsInfo.sourceTopics);
+            }
+        }
+        try {
```
nit: add empty line to separate blocks

k-apol replied:
Thank you, made an update. Resolving this

mjsax commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -293,6 +302,168 @@ public boolean isValidTransition(final State newState) {
+        try {
+            final ValidationResult validationResult = internalTopicManager.validate(allInternalTopics); // can throw timeout
+
+            final boolean noInternalTopicsExist = allInternalTopics.keySet() == validationResult.missingTopics();
```
```suggestion
            final boolean noInternalTopicsExist = allInternalTopics.keySet().equals(validationResult.missingTopics());
```

mjsax replied:
Using `==` we don't compare the content of both sets, but if both sets are literally the same object.

k-apol replied:
Updated, resolving this one

mjsax commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -293,6 +302,168 @@ public boolean isValidTransition(final State newState) {
+            throw new TimeoutException(timeoutException.getMessage(), timeoutException);
+        } catch (final StreamsException streamsException) {
+            throw new StreamsException(streamsException.getMessage(), streamsException);
+        }
```
Not sure why we want to catch and rethrow? Seems we don't add any value doing this, but we can remove the complete try-catch block and let any exception just bubble up unmodified?

k-apol replied:
You're right, this is bad style - I will update this

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ChangelogTopics.java:
```diff
@@ -135,4 +135,4 @@ public Set<TaskId> statefulTaskIds() {
     }
-}
\ No newline at end of file
+}
```
nit: avoid unnecessary format changes (seems to be a difference in EOF encoding)

k-apol replied:
Thank you, updated this

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -628,6 +624,14 @@ protected Map<String, List<TopicPartitionInfo>> getTopicPartitionInfo(final Set<
+        this.isInitializing = true;
+        this.initTimeout = timeoutMs;
+    }
+    private boolean isInitializing() {
```
Seems this method is only called once, and it's rather trivial. We should remove it, as it does not add any value IMHO

mjsax commented on streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java:
```diff
@@ -481,7 +481,12 @@ public void shouldOverrideStreamsDefaultProducerConfigs() {
         assertEquals("10000", producerConfigs.get(ProducerConfig.LINGER_MS_CONFIG));
         assertEquals("30000", producerConfigs.get(ProducerConfig.TRANSACTION_TIMEOUT_CONFIG));
     }
-
```
nit: keep blank line between methods (also below: add blank line after this method)

k-apol replied:
Thank you, added this and resolving

mjsax commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -425,7 +431,7 @@ public void shouldCloseStartupTasksAfterFirstRebalance() throws Exception {
         prepareThreadState(streamThreadTwo, state2);
         try (final MockedConstruction<StateDirectory> constructed = mockConstruction(StateDirectory.class,
-            (mock, context) -> when(mock.initializeProcessId()).thenReturn(UUID.randomUUID()))) {
+                (mock, context) -> when(mock.initializeProcessId()).thenReturn(UUID.randomUUID()))) {
```
avoid unnecessary reformatting (might be an IDE setting that you need to adjust, or maybe even better, disable auto-reformatting entirely); same on many places below.

k-apol replied:
Thanks, found out it was a Java-specific setting in my IDE. Resolving this one

mjsax commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -1839,6 +1845,145 @@ public Uuid get(final long timeout, final TimeUnit timeUnit) {
         assertThat(didAssertThreadTwo.get(), equalTo(true));
         assertThat(didAssertGlobalThread.get(), equalTo(true));
     }
+    @Test
```
nit: add missing blank line

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -555,21 +560,12 @@ public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) {
-                }
-                log.info(
-                    "Topics {} could not be made ready. Will retry in {} milliseconds. Remaining time in milliseconds: {}",
+                maybeThrowTimeoutExceptionDuringMakeReady(
```
Why did you extract the code into this new helper? Seems it's only called once, so unclear to me what the benefit is compared to just inlined code?

k-apol replied:
When I tried adding it as inline, the code style rule for cyclomatic complexity broke. In order to make use of a timeout passed to `init(final Duration timeout)`, I had to add some more logical branches to check for which kind of timeout had happened during `makeReady()`

mjsax replied:
Oh. Our old friend. I see. Makes sense.

mjsax replied:
Thinking about the structor of the code, I would propose to pull the `if (!topicsNotReady.isEmpty()))` check into the new help method, and rename the method to `verifyAllTopicsCreated()` or something like this?

k-apol replied:
I like this idea, let me take a crack at this. Based on your other comments about the complexity of this method, I want to make sure I do not overlook similar opportunities for abstracting a bit here.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -555,21 +560,12 @@ public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) {
             }
 
             if (!topicsNotReady.isEmpty()) {
```
General comment: I am missing an update to `InternalTopicManager` to avoid creating new topics for the case that the feature is enabled, and topic are missing. For this case, we should not attempt to create any topic, but fail with an error right away.

k-apol replied:
I was planning on doing this as a separate PR, but reading this is makes more sense to include it on this one. Working on an addition for this.

k-apol replied:
@mjsax Just pushed a change that reads the config and throws if manual setup is enabled. Open to feedback on the approach I took

mjsax replied:
It's always a good idea to split out partial PRs. If you wand to go this way, we should work on such a partial PR first.

This PR does add the public `init()` methods and should "complete" the feature. Merging this PR w/o all code in place is kinda "risky", as it puts us into a weird partial state for the feature.

If we split out a PR that only adds the new config, and the check we just discuss, it's would be much cleaner. For now, we can just move forward as you already added the code in this PR, but if this PR gets too big, we could still split out some code, merge the other PR first and rebase this PR later to reduce it's size.

k-apol replied:
I see what you're saying - some of this code is mostly harmless if it's half baked (configs, maybe the new errors), but the init method could be fairly destructive if it's not functionally complete. 

I will keep that in mind as we move forward and for next time.

mjsax reviewed (commented):
Made an high-level initial pass.

It would be good to either add a new integration test, or re-use an existing one to test this feature end-to-end.

k-apol committed:
Allow errors to bubble up in init()
Update javadoc for init()
Misc cleanup

k-apol committed:
cleanup formatting

k-apol committed:
refactor makeReady to throw if manual setup enabled and not called by init method

k-apol committed:
refactor makeReady to throw if manual setup enabled and not called by init method.

Add descriptive helper to throw on invalid invoke

k-apol commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -569,7 +579,7 @@ public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) {
         log.debug("Completed validating internal topics and created {}", newlyCreatedTopics);
-
+        this.isInitializing = false;
         return newlyCreatedTopics;
```
Edit: Removed this in recent change. Per @mjsax , I got rid of the state introduced here and just created an overload that will be called from init

k-apol commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -90,6 +92,8 @@ public InternalTopicManager(final Time time,
         this.adminClient = adminClient;
+        this.isManualInternalTopicConfig = streamsConfig.getString(StreamsConfig.INTERNAL_TOPIC_SETUP_CONFIG)
+                                                        .equals(StreamsConfig.INTERNAL_TOPIC_SETUP_MANUAL);
 
```
There may be an opportunity to use the InitParameters class here, but this approach seemed a bit cleaner to me.

mjsax replied:
I would prefer to "flip" the check:
```
this.isManualInternalTopicConfig = StreamsConfig.INTERNAL_TOPIC_SETUP_MANUAL.equals(
                                       streamsConfig.getString(StreamsConfig.INTERNAL_TOPIC_SETUP_CONFIG)
                                   );
```

The advantage is, that we know that `INTERNAL_TOPIC_SETUP_MANUAL` is never `null`, so there no risk of a NPE. On the other hand, if `streamsConfig.getString(StreamsConfig.INTERNAL_TOPIC_SETUP_CONFIG)` would return `null` (what we still guard against, but this guard is weaker) we cannot call `equals()` on the `null` reference and crash.

As a rule of thumb: I alway prefer to use `<constant>.equals(<variable>)` instead of `<variable>.equals(<constant>)` to guard against NPE.

k-apol replied:
So the way you wrote this does functionally the same thing, but I'm introducing some risk that can be  avoided by just changing the order?

I think I understand, the constant is always going to be a reference to a string value of some kind, even if we change the string later. But it's possible that if someone isn't using this feature, or it's not defaulted etc. that a crash is caused without good reason?

mjsax replied:
> So the way you wrote this does functionally the same thing, but I'm introducing some risk that can be avoided by just changing the order?

Correct.

> But it's possible that if someone isn't using this feature, or it's not defaulted etc. that a crash is caused without good reason?

Yes. For this particular case, it does not matter too much, but in general, there is always a change that `<variable> == null` so you it's just best practice to not use `<variable>.equals(...)` w/o a `null` check for `<variable>` (by flipping to `<constant>.equals(...)` we don't need the `null` check and are safe.)

mjsax commented:
One more side comments: if we want to reduce the size of this PR, we could also do a separate PR that only adds the new exception classes. It would just add some "dead code" until we use them, so it's zero risk.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -461,15 +469,22 @@ public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) {
             final Set<String> tempUnknownTopics = new HashSet<>();
             topicsNotReady = validateTopics(topicsNotReady, topics, tempUnknownTopics);
+
+            if (this.isManualInternalTopicConfig && !this.isInitializing) {
```
nit: remove unnecessary `this.` (it's a general code style we agreed on for Kafka code base in general).

mjsax replied:
Is this check actually correct here? `validateTopics()` above would return all topics that we need to create, so if `topicsNotReady` is empty here, all topic we did expect exist and we should not throw an exception. Should it be(?):
```
if (!topicsNotReady.isEmpty() && isManualInternalTopicConfig) {
    throw new...
}
```
(for this case, we would also move the check further below inside the already existing block that tried to creates internal topics L490):
```
if (!topicsNotReady.isEmpty()) {
    if (isManualInternalTopicConfig) {
        throw new ...
    }

    final Set<NewTopic> newTopics = new HashSet<>();
```

k-apol replied:
> Is this check actually correct here? `validateTopics()` above would return all topics that we need to create, so if `topicsNotReady` is empty here, all topic we did expect exist and we should not throw an exception. Should it be(?):
> 
> ```
> if (!topicsNotReady.isEmpty() && isManualInternalTopicConfig) {
>     throw new...
> }
> ```
> 
> (for this case, we would also move the check further below inside the already existing block that tried to creates internal topics L490):
> 
> ```
> if (!topicsNotReady.isEmpty()) {
>     if (isManualInternalTopicConfig) {
>         throw new ...
>     }
> 
>     final Set<NewTopic> newTopics = new HashSet<>();
> ```

My thinking was this, 

if `(!topicsNotReady.isEmpty())` evaluates to true (the break condition for the while loop), we are saying **there are some topics to create**.. but, if I understand your point, that decision is not actually determined until validate finishes executing?

Validate could return a result that says '**there's nothing to create', ** but the way this check stands, it would fail prematurely. If we decide to fail, we would need to be certain that there is a need to actually create topics to reach the desired state. I think capturing this decision in a variable could make this more readable in a future refactor.

 I will read over this and think on it, I think you may be right

mjsax replied:
> if (!topicsNotReady.isEmpty()) evaluates to true (the break condition for the while loop), we are saying there are some topics to create.. but, if I understand your point, that decision is not actually determined until validate finishes executing?

Correct. When we enter this method for the very first time and reach the `while`-loop, `topicsNotReady` contains _all_ internal topics that we may need to create. Only after the first check which topics already exist, we get the actual set of topics that we still need to create.

> Validate could return a result that says '**there's nothing to create', ** but the way this check stands, it would fail prematurely

Exactly.

k-apol replied:
Removed unnecessary "this" references, updated the logical check as well. I had to break out another helper method, for now I will refactor `makeReady` as little as possible and follow-up with another refactoring PR to cleanup later. In hindsight, this would have been better for me to do first.

Resolving this, I feel we are on the same page with this

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -74,11 +76,14 @@ public class InternalTopicManager {
     private final long retryBackOffMs;
     private final long retryTimeoutMs;
+    private Duration initTimeout;
+    private boolean isInitializing = false;
```
Not sure if I understand the semantics of this variable and why we need it. Can you explain?

k-apol replied:
The problem I am trying to solve with the **initTimeout** variable (and most of the related code) is that the `InternalTopicManager`, where most of the work happens, has it's own timeout exception that it can throw. I wanted a way to differentiate between the existing case and the case where the user passes a timeout into one of the overloads. 

Since the bulk of the work done by the init() method actually happens on the `InternalTopicManager`. It made sense to me, if the user wants to call one of the timeout overloads for the init constructor, to pass that timeout into the `InternalTopicManager.` `makeReady()` has it's own paths where a timeout exception can be thrown (the processing takes longer than the configured `MAX_POLL_INTERVAL_MS_CONFIG)`, I took this approach to be able to let the method throw a timeout exception in both cases. There may be a better way to do this, I also thought of doing something similar to Javascript's [promise.race](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/race) command. 

The current approach adds a lot of other surrounding code, and does not accurately represent the timeout a user would pass anyway. It's not accounting for any time that the validation takes before `makeReady` might be called. How would you approach it @mjsax ?

k-apol replied:
The problem I am trying to solve with the **isInitializing** flag is that, as I understand from the KIP, we want to prevent makeReady from being invoked by anything besides init when configured to do so. My idea was to set this flag is set to true when the init method is used to invoke `makeReady`, and `makeReady` turns it off at the end of it's work. This way, we can stop topics from being created if `makeReady` is called explicitly, instead of via `init` when applicable. Currently, it's tied to the timeout setting that also looks like it's going to be refactored somehow. There are other ways to stop topics from being created conditionally, and this was the simplest approach I could think of. 

This is where I currently try to decide whether or not topic creation is valid, just before we actually start creating them on line 488
![image](https://github.com/user-attachments/assets/5d361baf-c014-4a3a-956b-e723840331a9)

mjsax replied:
Ah, I see. Might be simpler to just add a `boolean` parameter to `makeReady` and pass in `true` from `init()` and `false` from the other calls to it?

In the end, we don't need to track this as "state" (ie, member variable on `InternalTopicManager`).

k-apol replied:
Pushed this up in [recent change](https://github.com/apache/kafka/pull/19913/commits/b11bb29933de7416f7b61f5037da435d1748c5a0).

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -461,15 +469,22 @@ public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) {
+
             newlyCreatedTopics.addAll(topicsNotReady);
 
             if (!topicsNotReady.isEmpty()) {
```
To reduce complexity of this class, we should also extract this block into a sub-method (maybe in a separate refactoring PR to not convolute PRs.) called `createTopics()` or similar?

k-apol replied:
I think it's a good idea, we are making this method work really hard. He could use some pals to help him out

k-apol replied:
@mjsax I have a small PR open to add just the exceptions, happy to open another small one to refactor this method and make it easier to work with as part of this PR. Let me know what you think, can always do this after as well. Trying to avoid as much refactoring as possible in this PR, despite that there is a lot needed and still a bit more to come.

mjsax replied:
Yes, sounds like a good idea. It's way easier to me to review, and should allow us to move forward is small incremental steps, end shorted end-to-end latency to complete the work.

k-apol commented:
> One more side comments: if we want to reduce the size of this PR, we could also do a separate PR that only adds the new exception classes. It would just add some "dead code" until we use them, so it's zero risk.

I don't mind doing this - if you think it's a good idea, it would be less painful to do it now than later. I feel that the current state of this PR is fairly volatile. You would have better judgment than I would on what constitutes a large PR for Kafka

**Update 6/18**: @mjsax I've just opened a [PR](https://github.com/apache/kafka/pull/19988) that does this. Maybe I can follow it with the refactoring changes we've outlined in the InternalTopicManager? Many of the changes there are tied to the new init method, but I could still 'break out' parts of the code to make it easier to integrate with init.

k-apol committed:
refactor makeReady to throw if manual setup enabled and not called by init method.

Add descriptive helper to throw on invalid invoke

k-apol committed:
refactor makeReady to throw if manual setup enabled and not called by init method.

Add descriptive helper to throw on invalid invoke

Update logical check for isManualInternalTopicConfig to avoid NPE

k-apol committed:
Merge remote-tracking branch 'origin/KAFKA-10357-Accidental-Repartition-Handling' into KAFKA-10357-Accidental-Repartition-Handling

# Conflicts:
#	streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java

mjsax commented:
Seems there is still a few smaller comments to address.

Just let me know when you need another round of review.

k-apol committed:
Merge branch 'apache:trunk' into KAFKA-10357-Accidental-Repartition-Handling

k-apol committed:
Add formatting updates - empty lines between blocks, remove uneccessary line break in ChangelogTOpics.java

k-apol committed:
formatting updates - remove excess indent

k-apol committed:
formatting - missing blank line

k-apol committed:
Code style update: remove unneccessary "this" references in InternalTopicManager.java

k-apol committed:
Update logical check to throw once topic creation starts. See: https://github.com/apache/kafka/pull/19913/#discussion_r2146261373

k-apol committed:
Update tests to match new init structure

k-apol committed:
decouple logical checks - these should be separate questions, not an else-if

k-apol committed:
remove unnecessary this ref

k-apol committed:
remove unnecessary this ref

k-apol committed:
formatting update

k-apol committed:
refactor away from state-based initialization check, rather provide an overload to do this ad-hoc - see https://github.com/apache/kafka/pull/19913#discussion_r2162812021

k-apol committed:
update makeReady call from init method

k-apol committed:
formatting update

mjsax commented:
Blocked on some other PRs, that we split out (just commenting so the PR does not get auto-closed).
